keps/sig-api-machinery/4355-coordinated-leader-election/README.md

```diff
@@ -0,0 +1,1377 @@
+
+### Enabling on a component
+
+Components with a `--leader-elect-resource-lock` flag (kube-controller-manager,
+ kube-scheduler) will accept `coordinatedleases` as a resource lock type.
+
+### Migrations
+
+So long as the API server is running a coordinated election controller, it is
+safe to directly migrate a component from Lease Based Leader Election to
+Coordinated Leader Election (or vis-versa).
+
+During the upgrade, a mix of components will be running both election
+approaches. When the leader lease expires, there are a couple possibilities:
+
+- A controller instance using `Lease`-based leader election claims the leader
+  lease
+- The coordinated election controller picks a leader, from the components that
+  have written LeaseCandidate leases, and claims the lease on the leader's behalf
+
+Both possibilities have acceptable outcomes during the migration: a component
+is elected leader, and once elected, remains leader so long as it keeps the
+lease renewed. The elected leader might not be the leader that Coordinated
+Leader Election would pick, but this is no worse than how leader election works
+before the upgrade, and once the upgrade is complete, Coordinated Leader
+Election works as intended.
+
+There is one thing that could make migrations slightly cleaner: If Coordinated
+Leader Election adds a `coordination.k8s.io/elected-by:
+leader-election-controller` annotation to any leases that it claims. It can also
+check for this annotation and only mark leases as "end-of-term" if that
+annotation is present. Lease Based Leader Election would ignore "end-of-term"
+annotations anyway, so this isn't strictly needed, but it would reduce writes
+from the coordinated election controller to leases that were claimed by
+component instances not using Coordinated Leader Election
+
+### API
+
+The lease lock API will be extended with a new field for election preference, denoted as an enum for strategies for Coordinated Leader Election.
+
+```go
+
+type CoordinatedLeaseStrategy string
+
+// CoordinatedLeaseStrategy defines the strategy for picking the leader for coordinated leader election.
+const (
+  OldestCompatibilityVersion CoordinatedStrategy = "OldestCompatibilityVersion"
+  NoCoordination CoordinatedStrategy = "NoCoordination"
+)
+
+type LeaseSpec struct {
+	// Strategy indicates the strategy for picking the leader for coordinated leader election
+  // This is filled in from LeaseCandidate.Spec.Strategy or defaulted to NoCoordinationStrategy
+  // if the leader was not elected by the CLE controller.
+	Strategy CoordinatedLeaseStrategy `json:"strategy,omitempty" protobuf:"string,6,opt,name=strategy"`
+
+	// EndofTerm signals to a lease holder that the lease should not be
+	// renewed because a better candidate is available.
+	EndOfTerm bool `json:"endofterm,omitempty" protobuf:"boolean,7,opt,name=endofterm"`
```

Comment:
```suggestion
	EndOfTerm bool `json:"endOfTerm,omitempty" protobuf:"boolean,7,opt,name=endOfTerm"`
```
Note that booleans are discouraged in APIs. We likely have to think about a string enum.

Reply:
Fixed casing. Can you elaborate on why bool is discouraged?